go/vt/vttablet/tabletmanager/vreplication/vcopier.go

```diff
@@ -364,3 +631,562 @@ func (vc *vcopier) fastForward(ctx context.Context, copyState map[string]*sqltyp
+//
+//	if _, err := task.Lifecycle().Before(nextState).Notify(ctx, args); err != nil {
+//		return err
+//	}
+//	if _, err := fn(ctx, args); err != nil {
+//		return err
+//	}
+//	if _, err := task.Lifecycle().After(nextState).Notify(ctx, args); err != nil {
+//		return err
+//	}
+//
+// Is roughly equivalent to:
+//
+//	if _, err := task.Lifecycle.TryAdvance(ctx, args, nextState, fn); err != nil {
+//	  return err
+//	}
+func (vtl *vcopierCopyTaskLifecycle) TryAdvance(
+	ctx context.Context,
+	args *vcopierCopyTaskArgs,
+	nextState vcopierCopyTaskState,
+	fn func(context.Context, *vcopierCopyTaskArgs) error,
+) (newState vcopierCopyTaskState, err error) {
+	// Canceling the task when there is a context error seems like a sensible
+	// thing to do. E.g. when copyPhaseDuration elapses it will cause the task
+	// to be canceled. Hower this is more aggressive cancelation behavior than
+	// the unit tests currently support. Will need to rework unit tests before
+	// we enable this.
+
+	//defer func() {
+	//	if err != nil {
+	//		//if errors.Is(err, context.Canceled) ||
+	//		//	errors.Is(err, context.DeadlineExceeded) {
+	//		//	newState = vcopierCopyTaskCancel
+	//		//} else {
+	//		//	newState = vcopierCopyTaskFail
+	//		//}
+	//	}
+	//}()
+
+	newState = nextState
+
+	// Checking for context errors seems like a sensible thing to do. E.g. when
+	// copyPhaseDuration elapses it will cause the task to be canceled. Hower
+	// this is more aggressive cancelation behavior than the unit tests
+	// currently support. Will need to rework unit tests before we enable this.
+
+	//if err = ctx.Err(); err != nil {
+	//	return
+	//}
+	if err = vtl.Before(nextState).Notify(ctx, args); err != nil {
+		newState = vcopierCopyTaskFail
+		return
+	}
+	if err = fn(ctx, args); err != nil {
+		newState = vcopierCopyTaskFail
+		return
+	}
+	if err = vtl.After(nextState).Notify(ctx, args); err != nil {
+		newState = vcopierCopyTaskFail
+		return
+	}
+	return
+}
+
+// Do registers a callback with the vcopierCopyTaskResultHooks, to be triggered
+// when a task advances to a "done" state (= canceled, completed, failed).
+func (vrh *vcopierCopyTaskResultHooks) Do(fn func(context.Context, *vcopierCopyTaskResult)) {
+	vrh.fns = append(vrh.fns, fn)
+}
+
+// Notify triggers all callbacks registered with this vcopierCopyTaskResultHooks.
+func (vrh *vcopierCopyTaskResultHooks) Notify(ctx context.Context, result *vcopierCopyTaskResult) {
+	for _, fn := range vrh.fns {
+		fn(ctx, result)
+	}
+}
+
+// SendTo registers a hook that accepts a result and sends the result to the
+// provided channel. E.g.:
+//
+//	resultCh := make(chan *vcopierCopyTaskResult, 1)
+//	task.Lifecycle().OnResult().SendTo(resultCh)
+//	defer func() {
+//	  result := <-resultCh
+//	}()
+func (vrh *vcopierCopyTaskResultHooks) SendTo(ch chan<- *vcopierCopyTaskResult) {
+	vrh.Do(func(ctx context.Context, result *vcopierCopyTaskResult) {
+		select {
+		case ch <- result:
+		case <-ctx.Done():
+			// Log a warning?
```

Comment:
What does this signify? We could block indefinitely on the channel receive, right? We failed to perform the requested task of sending the result to the channel, so shouldn't we return an error? We may even want to back all the way out, or no? What's the worst possible consequence of this?

Reply:
Right now we only use `SendTo` to do two things:

 * Notify the parent `copyTable` scope of completed work, by sending to the `resultCh` declared in that scope.
 * Notify the "next" copy task that it can proceed with updating `_vt.copy_state`.

In terms of what it signifies that `SendTo` would fail to send a result to `ch`, it could mean two things.

One: the parent `copyTable` scope isn't consuming completed work fast enough so `resultCh` fills up. If we get a context expiration while in that state, then `copyTable` scope could in theory close out the work pool, drain `resultCh`, but end up missing this result. If this result completed successfully, that would only impact stats and logging. But if it had an error, that would be bad. I need to think about this.

Two: this task can't notify the "next" task to proceed with `_vt.copy_state`. If we get a context expiration in that state, then the next task will also get a context expiration. That should be fine.

Reply:
From Zoom discussion:
 * In the current code in `main`, a context deadline/cancelation is handled by logging `Copy stopped` and by returning `nil` in `copyTable`.
 * The current async code in this branch works the same way, so that's good. We can improve things by trying to make it more clear that it works that way by code comments and maybe a bit of refactoring.